src/mlpack/core/tree/binary_space_tree/ub_tree_split_impl.hpp

```diff
@@ -0,0 +1,326 @@
+
+    for (; row < data.n_rows; row++)
+      if (lo[row] != hi[row])
+        break;
+
+    size_t bit = 0;
+
+    for (; bit < order; bit++)
+      if ((lo[row] & ((AddressElemType) 1 << (order - 1 - bit))) !=
+          (hi[row] & ((AddressElemType) 1 << (order - 1 - bit))))
+        break;
+
+    bit++;
+
+    // Replace insignificant bits.
+    if (bit == order)
+    {
+      bit = 0;
+      row++;
+    }
+    else
+    {
+      for (; bit < order; bit++)
+        lo[row] |= ((AddressElemType) 1 << (order - 1 - bit));
+      row++;
+    }
+
+    for (; row < data.n_rows; row++)
+      for (; bit < order; bit++)
+        lo[row] |= ((AddressElemType) 1 << (order - 1 - bit)); 
+  }
+
+  // The bound shouldn't contain too many subrectangles.
+  // In order to minimize the number of hyperrectangles we set last bits
+  // of the first address in the next node to 0 and last bits of the last
+  // address in the previous node to 1 in such a way that the ordering is not
+  // disturbed.
+  if (begin > 0)
+  {
+    // Omit leading equal bits.
+    size_t row = 0;
+    const arma::Col<AddressElemType>& lo = addresses[begin - 1].first;
+    arma::Col<AddressElemType>& hi = addresses[begin].first;
+
+    for (; row < data.n_rows; row++)
+      if (lo[row] != hi[row])
+        break;
+
+    size_t bit = 0;
+
+    for (; bit < order; bit++)
+      if ((lo[row] & ((AddressElemType) 1 << (order - 1 - bit))) !=
+          (hi[row] & ((AddressElemType) 1 << (order - 1 - bit))))
+        break;
+
+    bit++;
+
+    // Replace insignificant bits.
+    if (bit == order)
+    {
+      bit = 0;
+      row++;
+    }
+    else
+    {
+      for (; bit < order; bit++)
+        hi[row] &= ~((AddressElemType) 1 << (order - 1 - bit));
+      row++;
+    }
+
+    for (; row < data.n_rows; row++)
+      for (; bit < order; bit++)
+        hi[row] &= ~((AddressElemType) 1 << (order - 1 - bit)); 
+  }
+
+  // Set the minimum and the maximum addresses.
+  for (size_t k = 0; k < bound.Dim(); k++)
+  {
+    bound.LoAddress()[k] = addresses[begin].first[k];
+    bound.HiAddress()[k] = addresses[begin + count - 1].first[k];
+  }
+  bound.UpdateAddressBounds();
+
+  // Since the dataset is sorted we can easily get the split column.
+  splitCol = begin + count / 2;
+  
+  return true;
+}
+
+template<typename BoundType, typename MatType>
+bool UBTreeSplit<BoundType, MatType>::SplitNode(BoundType& bound,
+                                                MatType& data,
+                                                const size_t begin,
+                                                const size_t count,
+                                                size_t& splitCol,
+                                                std::vector<size_t>& oldFromNew)
+{
+  constexpr size_t order = sizeof(AddressElemType) * CHAR_BIT;
+  if (begin == 0 && count == data.n_cols)
+  {
+    InitializeAddresses(data);
+
+    std::sort(addresses.begin(), addresses.end(),ComparePair);
```

Comment:
This is a minor issue; there is a space missing between `addresses.end(),` and `ComparePair`. :)